### PR TITLE
Always Present Checkout button.

### PR DESCRIPTION
**Pull request by pschutz93**
# Description
Introducing the always present checkout button into the M2 code base. The button appears and enables initiation of checkout from any white listed when there are one or more items in the cart. The feature exists behind a feature switch. We respect the Bolt Everywhere flag for minicart and appears on whitelisted pages (the same as minicart). We likely will disable minicart for merchants who use this button but the code should still support minicart and always present checkout button on the same page.

image

The button is:

Hidden when no items in cart.
When an item is added to cart the Buy Now button appears. Minicart animates.
When navigating to another product or category page the Buy Now button appears.
To setup the button call BoltCheckout.configure with {floatingButtonOption: "show_cart"} or {floatingButtonOption: "show_cart_on_hover"} as the last argument depending on whether an item was just added or has been there.

To call when an item is added to cart, I set a listener on "ajax:addToCart". I call createOrder with dataChanged set to true which triggers configure with {floatingButtonOption: "show_cart"}`.

The tricky part is determining whether an item was just added to cart, has been there, or if there are no items in the cart. I check to see if there are now items in the cart by modifying monitorDataChange to watch for cart counter going from empty to a number. When this occurs we call createOrder call which triggers BoltCheckout.configure(cart, hints, callbacks, {floatingButtonOption: "show_cart_on_hover"}).

Another tricky part is not calling configure from monitorDataChange when we already called from the "ajax:addToCart" listener. We do this by checking what changed in monitorDataChange to only capture the cart being updated on page load (transitioning from empty string to a number).

See Product Spec here: https://docs.google.com/document/d/1EoTBhNeRPAj-RupFczAaLHakYgqonYMJvAjMFIm-EFk/edit#heading=h.dfj49q9kev35

See Technical Spec here: 
https://docs.google.com/document/d/1wP962zqqvyBPWQx7F3yXA48a8BtX1lN1C5Ni0hVqncY/edit

Fulfills:
https://app.asana.com/0/1146941972725303/1165704788287983/f
https://app.asana.com/0/1146941972725303/1176822385483791/f
https://app.asana.com/0/1146941972725303/1176822385483793/f

#changelog Always Present Checkout button.
Adds support for Always Present Checkout button. Adds feature switch for button.

# Type of change

- [ ] Bug fix (change which fixes an issue)
- [X] New feature (change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?
Please validate that you have tested your change in at least one of the following areas:

- [X] Successfully tested locally (or docker image)
- [ ] Successfully tested on a staging or sandbox server
- [ ] Successfully tested on a merchant's staging server

# Checklist:

- [X] My code follows the style guidelines of this project.
- [X] I have performed a self-review of my own code.
- [X] I have commented my code, particularly in hard-to-understand areas.
- [X] New and existing unit tests pass locally with my changes.
- [ ] I have created or modified unit tests to sufficiently cover my changes.
	- Added unit tests for feature switch. JS changes not super unit testable.
- [X] I have added my Jira ticket link and provided a changelog message.

### CODE REVIEW

File: Block/Js.php [hunk 1]
```diff
@@ -226,6 +226,7 @@ public function getSettings()
             'is_pre_auth'              => $this->getIsPreAuth(),
             'default_error_message'    => $this->getBoltPopupErrorMessage(),
             'button_css_styles'        => $this->getButtonCssStyles(),
+            'always_present_checkout' => $this->featureSwitches->isAlwaysPresentCheckoutEnabled(),
```
Comment (matt-thomason) [inline Block/Js.php:229]: You've used feature switches here which is good, but something like this should also be controllable by the merchant in their plugin settings. Look at `pay_by_link_url` a few lines above for an example. Can you add that behaviour as well so it's toggle-able in M2?

**Comment by matt-thomason** [comment]
There also appear to be conflicts - please do a rebase @pschutz93 :)

File: Block/Js.php (hunk 1 above)
Comment (pschutz93) [reply Block/Js.php:229]: To do this do I just add a config path and call something like `$this->getCustomURLValueOrDefault(self::ALWAYS_PRESENT_CHECKOUT_CONFIG_PATH, false);` and define `ALWAYS_PRESENT_CHECKOUT_CONFIG_PATH = "payments/boltpay/always_present_checkout"`? Is there anywhere to define a toggle-able switch that shows up in setings panel?

Comment (matt-thomason) [reply Block/Js.php:229]: Yes, in `etc/config.xml`.

Comment (matt-thomason) [reply Block/Js.php:229]: You also need it in `etc/adminhtml/system.xml`.

Comment (pschutz93) [reply Block/Js.php]: Fixed up

File: Helper/Config.php [hunk 2]
```diff
@@ -1706,6 +1712,20 @@ public function isStorePickupFeatureEnabled($storeId = null)
         );
     }
 
+    /**
+     * Get config value for whether or not always present checkout is enabled
+     * @param int|string|Store $store
+     * @return boolean
+     */
+    public function isAlwaysPresentCheckoutEnabled($store = null)
```
Comment (matt-thomason) [inline Helper/Config.php:1720]: Our convention here is `$storeId`

File: etc/adminhtml/system.xml [hunk 3]
```diff
@@ -95,6 +95,11 @@
                         <config_path>payment/boltpay/order_management</config_path>
                         <source_model>Magento\Config\Model\Config\Source\Yesno</source_model>
                     </field>
+                    <field id="always_present_checkout" translate="label" type="select" sortOrder="83" showInDefault="0" showInWebsite="0" showInStore="0">
+                        <label>(BETA) Enable Always Present Checkout Button</label>
```
Comment (matt-thomason) [inline etc/adminhtml/system.xml:99]: @njfaries An idle thought here - maybe we should consider having a section for developer options/beta features, similar to what we have in the admin dash. Then our beta features aren't in the middle of the regular config settings.

Out of scope for this PR but it had me thinking.

File: view/frontend/templates/js/replacejs.phtml [hunk 4]
```diff
@@ -806,8 +805,12 @@ ob_start();
                         createRequest = false;
                     })
             });
+            // Get number of items in cart from counter element
+            var numItems = $(".counter-number").text();
 
-            var BC = BoltCheckout.configure(cart, hintBarrier.promise, callbacks);
+            // Call config with floatingButtonMode if always present checkout enabled and items were or were not just added to cart (dataChanged)
+            var config = boltConfig.always_present_checkout && (numItems && numItems !=="0" || dataChanged)? {floatingButtonMode: dataChanged ? "show_cart" : "show_cart_on_hover" } : {};
+            var BC = BoltCheckout.configure(cart, hintBarrier.promise, callbacks, config);
```
Comment (vitaliyreznikov) [inline view/frontend/templates/js/replacejs.phtml:813]: We have three configure call in this file. Could you please add your code to two others as well? It's possible that numItems still 0 at this moment because data is still loading. Later we re-configure bolt, but from another place and don't calculate config.
I think we can just copy-paste the code. We are planning to refactor this JS code soon.

File: etc/adminhtml/system.xml [hunk 5]
```diff
@@ -95,6 +95,11 @@
                         <config_path>payment/boltpay/order_management</config_path>
                         <source_model>Magento\Config\Model\Config\Source\Yesno</source_model>
                     </field>
+                    <field id="always_present_checkout" translate="label" type="select" sortOrder="83" showInDefault="0" showInWebsite="0" showInStore="0">
```
Comment (vitaliyreznikov) [inline etc/adminhtml/system.xml:98]: Do you want to hide this setting?

File: view/frontend/templates/js/replacejs.phtml [hunk 6]
```diff
@@ -859,6 +862,9 @@ ob_start();
             checkout_buttons = true;
             processButtons();
         });
+        if (boltConfig.always_present_checkout) {
+            processButtons();
+        }
```
Comment (vitaliyreznikov) [inline view/frontend/templates/js/replacejs.phtml:867]: when we have in cart both simple and configurable product and click “add to cart” (I do it for the same product that is already on cart)
When we check .block-minicart .subtitle it related to two different DOM nodes: subtitle of minicart and subtitle of configurable product
Function monitorDataChange thinks this was update and call createOrder() again and again.

If we delete lines 865-867 then after few attempts this reloading is finished.
I propose to remove them and completely fix the issue later when we refactor this JS file.

File: Helper/FeatureSwitch/Definitions.php [hunk 7]
```diff
@@ -168,6 +173,12 @@ class Definitions {
             self::VAL_KEY             => true,
             self::DEFAULT_VAL_KEY     => false,
             self::ROLLOUT_KEY         => 0
+        ),
+        self::M2_ALWAYS_PRESENT_CHECKOUT => array(
+            self::NAME_KEY => self::M2_ALWAYS_PRESENT_CHECKOUT,
+            self::VAL_KEY => true,
+            self::DEFAULT_VAL_KEY => false,
+            self::ROLLOUT_KEY => 0,
```
[the remark this answers is not in the document]
Comment (pschutz93) [reply Helper/FeatureSwitch/Definitions.php:181]: I want to default rollout to off and require both config and bolt feature switch to be enabled. This is a brand new feature so it makes sense to have disabled by default even for merchants that turn on. We also at this time don't want merchants to be able to enable always present checkout on their own.

File: etc/adminhtml/system.xml (hunk 5 above)
Comment (pschutz93) [reply etc/adminhtml/system.xml:98]: I want to show it. This way a merchant who we enable through feature switch can also turn it off if they want.

File: view/frontend/templates/js/replacejs.phtml (hunk 4 above)
Comment (pschutz93) [reply view/frontend/templates/js/replacejs.phtml:814]: Updated. I only want to call with "show_cart" when a new item is added (when ajax:addToCart event is fired) so for other two places I call with only "show_cart_on_hover" and remove dataChanged param.

File: view/frontend/templates/js/replacejs.phtml (hunk 6 above)
Comment (pschutz93) [reply view/frontend/templates/js/replacejs.phtml]: Removed

File: view/frontend/templates/js/replacejs.phtml (hunk 4 above)
Comment (pschutz93) [reply view/frontend/templates/js/replacejs.phtml]: Actually we can't call everywhere. The issue is that we only want to call with config when a cart with items is created (which to my knowledge happens only in `createOrder`. If I call with config later `BC.configure` is called with a null cart. This breaks always present button as it depends on having cart to show. We only want to call it when passing a cart that is not null.

Comment (pschutz93) [reply view/frontend/templates/js/replacejs.phtml:814]: I also changed monitorChagnes function quite a bit which should make fix edge cases where number hasn't updated uet.

Comment (vitaliyreznikov) [reply view/frontend/templates/js/replacejs.phtml:814]: On my test site when I am on product page:
1) The first configure call happens on line 813. numItem is "0" at this moment, we don't show always present button
2) The second call happens on line 916. Cart is set and we can check $(".counter-number").text() and show button if we copy the same code before line 916.

So for now on my test site button is shown if I add product to cart, but isn't shown if reload the page.

File: etc/adminhtml/system.xml (hunk 5 above)
Comment (vitaliyreznikov) [reply etc/adminhtml/system.xml:98]: I see. Then we need set parameters to:
**showInDefault="1" showInWebsite="1" showInStore="0"**
